ZencashOfficial/arizen
# Secure nodes / Z Transactions

Implementing issue #302 

- [x] Secure Node settings
- [x] Arizen Settings
- [x] General RPC communication
- [x] RPC console (for testing)
- [x] RPC over SSH ~~HTTPS communication~~

Specific functions
- [x] Import PK in node (if not there)
- [x] New Z address Balance
- [x] Get Balance
- [x] Send ZEN function

Integrate to Arizen UI
- [x] New Z address
- [x] Update db
- [x] Update Separately T balance
- [x] Update Separately Z balance
- [x] Send From or To Z address GUI integration
- [x] Import Z address possible
~~List Transactions if possible~~

### User experience
- [x] Import all Z PKs from SN to Arizen
- [x] Import all Z PKs from Arizen to SN (this takes time as SN rescans)
~~Import all PKs from Arizen to Secure node, only if used to send txs~~
- [x] Export Spending Key for Z addr

### Error indications
- [x] SSH tunnel
- [x] zend is not working

### Improve
- [x]  Fix: Do not open too many SSH tunnels
- [x] Use intermediate address ~~import of T addressed in Super Node~~
- [x] Import of new Z PKs - "User should rescan" message

### More

### Known issues:
- [x] Solved: ~~Address string is 95 characters and destroys GUI. Should be displayed like zce...kjhuoC~~
- [x] Solved: ~~Current implementation of "Update Z balance" uses async update of balance and update of GUI. if GUI updates first old balance is displayed.~~
- [x] Transaction is placed as using the intermediate T~~Transaction id is not placed for txs, Z to T, T to Z (T to T works fine, Z to Z should be there ?)~~
- [x] Intermediate T is used ~~Arizen uses compressed T address while zen-cli cannot (?)~~

### To-Do later:
- [x] ~~Batch withdraw for Z addresses ?~~
- [x] ~~Add memo for to/from Z txs~~

### Instructions for Settings in Arizen
1. Secure Node FQDN = mynode.example.com
2. SSH username, password and port used (to access your VM)
3. if rpcport=8231 (in zen.conf), then Secure Node Port is 8231
3. Secure Node Username  (in zen.conf)
4. Secure Node Password  (in zen.conf)


### Testing
1. create a dummy SN (or use an existing)
2. start zend and wait for sync
3. Open arizen and fill in the settings
4. The leds should update to green
5. Open RPC test (console) and try the help command. You should get result. try any command.
6. Get a new Z address
7. send from another wallet 0.0001 zen and wait to see if balance updates automatically after the tx is confirmed.
8. After tx is confirmed send from your Z to another Z or T(to verify you can send ZEN from this address)
9. Verify that T to T works by sending (T to T)
10. from Console or your terminal create a new Z address (zen-cli z_getnewaddress) and wait for arizen to import this address from SN (after refresh)
11. make a Z address from a 3rd wallet and send zen. Import the PK in Arizen, and after refresh the Z address should be in your SN. Is the balance correct ?
12. power off your SN and see that leds become red
